File: lab8-k/scanner.py
from coords import Position, Fragment
from tok import Token, Tag


class Scanner:
    def __init__(self, input, compiler):
        self.text = input
        self.__compiler = compiler
        self.__cur = Position(self.text)
        self.__comments = []

    def get_comments(self):
        return self.__comments
    
    def add_comment(self, start, end):
        self.__comments.append(Fragment(start, end))

    def next_token(self):
        while self.__cur.cp():
            while self.__cur.is_white_space():
                self.__cur += 1
            start = self.__cur.copy()

            if self.__cur.cp() == '/':
                self.__cur += 1

                if self.__cur.cp() == '*':
                    self.__cur += 1
                    closed = False

                    while not closed:
                        while self.__cur.cp() != '*':
                            if self.__cur.is_eof():
                                self.__compiler.add_msg(False, self.__cur.copy(), 'Unexpected EOF')
                                self.__compiler.add_comment(start.copy(), self.__cur.copy())
                                return Token(Tag.COMMENT, start.copy(), self.__cur.copy())
                            self.__cur += 1
                        self.__cur += 1
                        if self.__cur.cp() == '/':
                            closed = True
                            self.__cur += 1

                    pos = self.__cur.copy()
                    self.add_comment(start.copy(), self.__cur.copy())
                    return Token(Tag.COMMENT, start.copy(), self.__cur.copy())
            elif self.__cur.cp() == '"':
                self.__cur += 1

                text = self.__cur.cp()
                self.__cur += 1

                if self.__cur.is_eof():
                    self.__compiler.add_msg(False, self.__cur.copy(), 'Unexpected EOF')
                    return Token(Tag.TERM, start.copy(), self.__cur.copy(), text)
                if self.__cur.cp() != '"':
                    self.__compiler.add_msg(False, self.__cur.copy(), 'Expected close bracket')
                else:
                    self.__cur += 1
                return Token(Tag.TERM, start.copy(), self.__cur.copy(), text)
            elif self.__cur.is_upper_case():
                text = self.__cur.cp()
                self.__cur += 1
                return Token(Tag.NTERM, start.copy(), self.__cur.copy(), text)
            elif self.__cur.cp() == '(':
                self.__cur += 1
                return Token(Tag.LPAREN, start.copy(), self.__cur.copy())
            elif self.__cur.cp() == ')':
                self.__cur += 1
                return Token(Tag.RPAREN, start.copy(), self.__cur.copy())
            elif self.__cur.cp() == '{':
                self.__cur += 1
                return Token(Tag.LFPAREN, start.copy(), self.__cur.copy())
            elif self.__cur.cp() == '}':
                self.__cur += 1
                return Token(Tag.RFPAREN, start.copy(), self.__cur.copy())
            elif self.__cur.cp() == '[':
                self.__cur += 1
                return Token(Tag.LSPAREN, start.copy(), self.__cur.copy())
            elif self.__cur.cp() == ']':
                self.__cur += 1
                return Token(Tag.RSPAREN, start.copy(), self.__cur.copy())
            elif self.__cur.cp() == ',':
                self.__cur += 1
                return Token(Tag.COMMA, start.copy(), self.__cur.copy())
            else:
                self.__compiler.add_msg(True, self.__cur.copy(), 'Unexpected symbol')
                self.__cur += 1
                break

        return Token(Tag.EOF, self.__cur.copy(), self.__cur.copy())

File: lab8-a/src/Scanner.java
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

public class Scanner {
    final public String text;
    private final Compiler compiler;
    private final Position cur;
    private final List<Fragment> comments;

    public Collection<Fragment> getComments() {
        return comments;
    }

    public Scanner(String text, Compiler compiler) {
        this.compiler = compiler;
        this.text = text;
        cur = new Position(this.text);
        comments = new ArrayList<>();
    }

    public Token nextToken() {
        while (cur.cp() != (char) -1) {
            while (cur.isWhiteSpace()) {
                cur.inc();
            }
            Position start = cur.clone();

            if (cur.cp() == '/') {
                cur.inc();
                if (cur.cp() == '*') {
                    cur.inc();
                    boolean isClosed = false;

                    while (!isClosed) {
                        while (cur.cp() != '*') {
                            if (cur.isEnd()) {
                                compiler.addMessage(false, cur.clone(), "Unexpected end of comment");
                                compiler.addComment(start.clone(), cur.clone());
                                return new Token(DomainTag.COMMENT, start.clone(), cur.clone(), 'N');
                            }
                            cur.inc();
                        }
                        cur.inc();
                        if (cur.cp() == '/') {
                            isClosed = true;
                            cur.inc();
                        }
                    }
                    compiler.addComment(start.clone(), cur.clone());
                    return new Token(DomainTag.COMMENT, start.clone(), cur.clone(), 'N');
                } else {
                    compiler.addMessage(true, cur.clone(), "Unexpected characted");
                }
            } else if (cur.cp() == '"') {
                cur.inc();

                char value = cur.cp();
                cur.inc();

                if (cur.isEnd()){
                    compiler.addMessage(false, cur.clone(), "Unexpected end of term");
                    return new Token(DomainTag.TERM, start.clone(), cur.clone(), value);
                }
                cur.inc();
                return new Token(DomainTag.TERM, start.clone(), cur.clone(), value);
            } else if (cur.isUpperCase()) {
                char value = cur.cp();
                cur.inc();

                return new Token(DomainTag.NTERM, start.clone(), cur.clone(), value);
            } else {
                switch (cur.cp()) {
                    case '(':
                        cur.inc();
                        return new Token(DomainTag.LPAREN, start.clone(), cur.clone(), 'N');
                    case ')':
                        cur.inc();
                        return new Token(DomainTag.RPAREN, start.clone(), cur.clone(), 'N');
                    case '{':
                        cur.inc();
                        return new Token(DomainTag.LFPAREN, start.clone(), cur.clone(), 'N');
                    case '}':
                        cur.inc();
                        return new Token(DomainTag.RFPAREN, start.clone(), cur.clone(), 'N');
                    case '[':
                        cur.inc();
                        return new Token(DomainTag.LSPAREN, start.clone(), cur.clone(), 'N');
                    case ']':
                        cur.inc();
                        return new Token(DomainTag.RSPAREN, start.clone(), cur.clone(), 'N');
                    case ',':
                        cur.inc();
                        return new Token(DomainTag.COMMA, start.clone(), cur.clone(), 'N');
                    default:
                        compiler.addMessage(true, cur.clone(), "Unexpected character");
                        cur.inc();
                        break;
                }
            }
        }

        return new Token(DomainTag.EOF, cur.clone(), cur.clone(), 'N');
    }
}
